presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java

```diff
@@ -148,9 +150,13 @@ private void writePages(Iterator<Page> pageIterator)
         try {
             InputStream input = new FileInputStream(targetFileName.toFile());
             memoryContext.setBytes(BUFFER_SIZE);
-            closer.register(input);
-            closer.register(() -> memoryContext.setBytes(0));
-            return PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE));
+            Iterator<Page> pages = PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE));
+            Closeable closeable = () -> {
+                input.close();
+                memoryContext.setBytes(0);
+            };
+            closer.register(closeable);
```

Comment:
You still leak if `PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE))` throws.

It should be

```
InputStream input = closer.register(new FileInputStream(targetFileName.toFile()));
memoryContext.setBytes(BUFFER_SIZE);
Iterator<Page> pages = PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE));
return PrestoIterators.closeWhenExhausted(pages, () -> {
 input.close();
memoryContext.setBytes(0);
});
```
maybe last line should be even

```
return PrestoIterators.closeWhenExhausted(pages, this);
```
?